sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -111,24 +109,34 @@ case class InsertIntoHadoopFsRelationCommand(
       outputPath = outputPath.toString,
       dynamicPartitionOverwrite = dynamicPartitionOverwrite)
 
-    val doInsertion = (mode, pathExists) match {
-      case (SaveMode.ErrorIfExists, true) =>
-        throw new AnalysisException(s"path $qualifiedOutputPath already exists.")
-      case (SaveMode.Overwrite, true) =>
-        if (ifPartitionNotExists && matchingPartitions.nonEmpty) {
-          false
-        } else if (dynamicPartitionOverwrite) {
-          // For dynamic partition overwrite, do not delete partition directories ahead.
-          true
+    val pathExists = () => fs.exists(qualifiedOutputPath)
+
+    val doInsertion = mode match {
```

Comment:
I think the only exceptional case is Append - it doesn't need to know about `pathExists`. 

Personally current pattern matching looks cleaner and concise, it might be better to exclude Append via if statement and keep current pattern matching (without Append) in else statement. And seems like pathExists even could be a local variable - I can't find the usage otherwise.